[ADF]  Azure Data Factory updated ARM template

markingmyname opened the pull request:
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)


- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

markingmyname commented:
@bmoore-msft 

Hi, Brian,

I am creating this template on behalf of @daperlov. Once it is complete then he will take over from there.

I will keep watch of this PR.

I tried to follow the "best practices" in the Cg as best as I could understand it. :)

Thank you.

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,196 +1,209 @@
-      "type": "securestring",
-      "metadata": {
-        "description": "Key for the Azure storage account."
+        "storageAccountKey": {
```
remove this since you'll use the resourceId to get the storageAccount key using listkeys() below

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,196 +1,209 @@
-        "description": "Name of the blob container in the Azure Storage account."
+        "blobContainer": {
+            "type": "string",
+            "defaultValue":"GET-PREREQ-blobContainer",
```
same as all the other PREREQ comments (and same for those below too)

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,196 +1,209 @@
-        "description": "The folder in the blob container that will hold the transformed data."
-      }
+    "variables":{
+        "factoryId":"[concat('Microsoft.DataFactory/factories/', parameters('dataFactoryName'))]"
```
remove concat and use resourceId

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,196 +1,209 @@
+                        ],
+                        "type":"AzureBlobStorage",
+                        "typeProperties":{  
+                            "connectionString":"[concat('DefaultEndpointsProtocol=https;AccountName=',parameters('storageAccountName'),';AccountKey=',parameters('storageAccountKey'))]"
```
use listKeys() here to get the storage key

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,196 +1,209 @@
+                            "connectionString":"[concat('DefaultEndpointsProtocol=https;AccountName=',parameters('storageAccountName'),';AccountKey=',parameters('storageAccountKey'))]"
+                        }
+                    },
+                    "dependsOn":[  
```
move all dependsOn above properties

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,196 +1,209 @@
+                                ]
+                            }
+                        ],
+                        "annotations":[  
```
remove all empty/null properties

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.parameters.json:
```diff
@@ -1,12 +1,9 @@
-    },    
-    "dataFactoryLocation": {
-      "value": "East US"
+      "value": "Gen-Unique"
```
remove this param since you have a defaultValue in the template

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,196 +1,209 @@
+            "apiVersion":"2018-06-01",
+            "location":"[parameters('location')]",
+            "identity":{  
+                "type":"SystemAssigned"
```
how is this used?

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,196 +1,209 @@
+                        "typeProperties":{
+                            "location":{
+                                "type":"AzureBlobStorageLocation",
+                                "fileName":"emp.txt",
```
does this file exist?

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.parameters.json:
```diff
@@ -15,18 +12,12 @@
     },
     "blobContainer": {
-      "value": "adftutorial"
+      "value": "GET-PREREQ-blobContainer"
```
this isn't created by the prereq template

markingmyname commented:
Made changes as requested and updated the template to have fewer parameters but build more objects, such as storage account and blob container.

also updated values for the parameter file

markingmyname commented:
@bmoore-msft 

Hopefully, round 2 is better formatted. :)

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

markingmyname commented:
@bmoore-msft 

Is this failing because I have a storage access key being built via the template?

The API version is no older than 2018-06-01 for ADF, unfortunately. 

thank you.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

markingmyname commented:
#sign-off

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,209 +1,193 @@
+		"factoryId": "[resourceId('Microsoft.DataFactory/factories/', parameters('dataFactoryName'))]",
+		"storageAccountId": "[resourceId('Microsoft.Storage/storageAccounts', parameters('storageAccountName'))]",
+		"storageLinkedService": "[concat(variables('factoryId'), '/linkedServices/ArmtemplateStorageLinkedService')]",
+		"datasetIn": "[concat(variables('factoryId'), '/datasets/ArmtemplateTestDatasetIn')]",
```
remove concat and use resourceId()

markingmyname replied:
When I do that, I get an error:

Deployment template validation failed: 'The template variable 'datasetIn' is not valid: Unable to evaluate template language function 'resourceId': function requires exactly one multi-segmented argument which must be resource type including resource provider namespace. Current function arguments '/subscriptions/ab39dbc6-61cd-47cb-832c-c2c7c55391e8/resourceGroups/ADFARMtest/providers/Microsoft.DataFactory/factories/datafactoryva7rlg4aofcs6,/datasets/ArmtemplateTestDatasetIn'. Please see https://aka.ms/arm-template-expressions/#resourceid for usage details.. Please see https://aka.ms/arm-template-expressions for usage details.'.

I've seen in other templates, having contact in the main variables is acceptable.

Can I ask why not with these?

markingmyname replied:
done. :)

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,209 +1,193 @@
+		"storageAccountId": "[resourceId('Microsoft.Storage/storageAccounts', parameters('storageAccountName'))]",
+		"storageLinkedService": "[concat(variables('factoryId'), '/linkedServices/ArmtemplateStorageLinkedService')]",
+		"datasetIn": "[concat(variables('factoryId'), '/datasets/ArmtemplateTestDatasetIn')]",
+		"datasetOut": "[concat(variables('factoryId'), '/datasets/ArmtemplateTestDatasetOut')]"
```
remove concat() and use resourceId

markingmyname replied:
When I do that, I get an error:

Deployment template validation failed: 'The template variable 'datasetOut' is not valid: Unable to evaluate template language function 'resourceId': function requires exactly one multi-segmented argument which must be resource type including resource provider namespace. Current function arguments '/subscriptions/ab39dbc6-61cd-47cb-832c-c2c7c55391e8/resourceGroups/ADFARMtest/providers/Microsoft.DataFactory/factories/datafactoryva7rlg4aofcs6,/datasets/ArmtemplateTestDatasetOut'. Please see https://aka.ms/arm-template-expressions/#resourceid for usage details.. Please see https://aka.ms/arm-template-expressions for usage details.'.

I've seen in other templates, having contact in the main variables is acceptable.

Can I ask why not with these?

markingmyname replied:
done. :)

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,209 +1,193 @@
+	"variables": {
+		"factoryId": "[resourceId('Microsoft.DataFactory/factories/', parameters('dataFactoryName'))]",
+		"storageAccountId": "[resourceId('Microsoft.Storage/storageAccounts', parameters('storageAccountName'))]",
+		"storageLinkedService": "[concat(variables('factoryId'), '/linkedServices/ArmtemplateStorageLinkedService')]",
```
remove concat() and use resourceId()

markingmyname replied:
When I do that, I get an error:

Deployment template validation failed: 'The template variable 'storageLinkedService' is not valid: Unable to evaluate template language function 'resourceId': function requires exactly one multi-segmented argument which must be resource type including resource provider namespace. Current function arguments '/subscriptions/ab39dbc6-61cd-47cb-832c-c2c7c55391e8/resourceGroups/ADFARMtest/providers/Microsoft.DataFactory/factories/datafactoryva7rlg4aofcs6,/linkedServices/ArmtemplateStorageLinkedService'. Please see https://aka.ms/arm-template-expressions/#resourceid for usage details.. Please see https://aka.ms/arm-template-expressions for usage details.'.

I've seen in other templates, having contact in the main variables is acceptable.

Can I ask why not with these?

bmoore-msft replied:
you need to set the segments properly:

"storageLinkedService": "[resourceId('Microsoft.DataFactory/factories/linkedServices',  parameters('dataFactoryName'), 'ArmtemplateStorageLinkedService')]"

Follow the same pattern with the others

markingmyname replied:
done. :)

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,209 +1,193 @@
+							"connectionString": "[concat('DefaultEndpointsProtocol=https;AccountName=',parameters('storageAccountName'),';AccountKey=',listKeys(variables('storageAccountId'), '2019-06-01').keys[0].value)]"
+						}
+					},
+					"dependsOn": [
```
move all dependsOn properties above the properties property

markingmyname replied:
done. :)

bmoore-msft commented on 101-data-factory-v2-blob-to-blob-copy/azuredeploy.json:
```diff
@@ -1,209 +1,193 @@
+					"apiVersion": "2018-06-01",
+            		"location": "[parameters('location')]",
+					"properties": {
+						"activities": [{
```
format the JSON (e.g. SHIFT+ALT+F in VS Code)

markingmyname replied:
I did not see this off-formating when checking VS Code.

bmoore-msft replied:
line 147 shows a [ and a { on the same line (this is visible in VS Code).

markingmyname replied:
I was focused on location being out of place.

Corrected. :)

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

markingmyname commented:
#sign-off

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.

markingmyname commented:
#sign-off

bmoore-msft commented:
@daperlov - check this PR for updates that may be needed to documentation that references this sample.
